File: javascript/nws_utils.js
// nws_utils.js
const { 
  Coordinates, Temperature, Wind, Precipitation, 
  WeatherCondition, ForecastPeriod, Forecast, 
  Observation, Alert 
} = require('./weather_models');

/**
 * Parse raw forecast data into a Forecast object.
 * @param {Object} forecastData - Raw forecast data from the API.
 * @returns {Forecast} Parsed forecast object.
 */
function parseForecast(forecastData) {
  const properties = forecastData.properties || {};
  let updated = properties.updated || "";
  if (updated) {
    updated = new Date(updated);
  }

  const periods = [];
  for (const periodData of properties.periods || []) {
    const startTime = new Date(periodData.startTime || "");
    const endTime = new Date(periodData.endTime || "");
    
    const temperature = new Temperature(
      periodData.temperature || 0,
      periodData.temperatureUnit || "F"
    );
    
    const windDirection = periodData.windDirection || "";
    const windSpeedStr = periodData.windSpeed || "0 mph";
    const windSpeed = parseInt(windSpeedStr.replace(/[^0-9]/g, ""));
    
    const wind = new Wind(
      windSpeed,
      windDirection,
      "mph"
    );
    
    const period = new ForecastPeriod(
      periodData.name || "",
      startTime,
      endTime,
      temperature,
      wind,
      periodData.shortForecast || "",
      periodData.detailedForecast || "",
      periodData.icon,
      periodData.probabilityOfPrecipitation?.value
    );
    
    periods.push(period);
  }
  
  return new Forecast(updated, periods);
}

/**
 * Parse raw observation data into an Observation object.
 * @param {Object} observationData - Raw observation data from the API.
 * @returns {Observation} Parsed observation object.
 */
function parseObservation(observationData) {
  const properties = observationData.properties || {};
  
  const station = properties.station || "";
  const timestampStr = properties.timestamp || "";
  const timestamp = timestampStr ? new Date(timestampStr) : new Date();
  
  let temperature = null;
  if ("temperature" in properties) {
    temperature = new Temperature(
      properties.temperature.value || 0,
      (properties.temperature.unitCode || "wmoUnit:degC").split(":").pop()
    );
  }
  
  let dewpoint = null;
  if ("dewpoint" in properties) {
    dewpoint = new Temperature(
      properties.dewpoint.value || 0,
      (properties.dewpoint.unitCode || "wmoUnit:degC").split(":").pop()
    );
  }
  
  const relativeHumidity = properties.relativeHumidity?.value;
  
  let wind = null;
  if ("windSpeed" in properties && "windDirection" in properties) {
    wind = new Wind(
      properties.windSpeed.value || 0,
      properties.windDirection.value || 0,
      (properties.windSpeed.unitCode || "wmoUnit:km_h-1").split(":").pop()
    );
  }
  
  const barometricPressure = properties.barometricPressure?.value;
  const visibility = properties.visibility?.value;
  const textDescription = properties.textDescription || "";
  
  let precipitationLastHour = null;
  if ("precipitationLastHour" in properties) {
    precipitationLastHour = new Precipitation(
      properties.precipitationLastHour.value || 0,
      null,
      (properties.precipitationLastHour.unitCode || "wmoUnit:mm").split(":").pop()
    );
  }
  
  return new Observation(
    station,
    timestamp,
    temperature,
    dewpoint,
    relativeHumidity,
    wind,
    barometricPressure,
    visibility,
    textDescription,
    precipitationLastHour
  );
}

/**
 * Parse raw alert data into an Alert object.
 * @param {Object} alertData - Raw alert data from the API.
 * @returns {Alert} Parsed alert object.
 */
function parseAlert(alertData) {
  const properties = alertData.properties || {};
  
  // Parse timestamps
  const sent = new Date(properties.sent || "");
  const effective = new Date(properties.effective || "");
  const expires = new Date(properties.expires || "");
  
  const onsetStr = properties.onset;
  const onset = onsetStr ? new Date(onsetStr) : null;
  
  const endsStr = properties.ends;
  const ends = endsStr ? new Date(endsStr) : null;
  
  // Parse affected zones
  const affectedZones = [];
  for (const zone of properties.affectedZones || []) {
    const zoneId = zone.split("/").pop();
    affectedZones.push(zoneId);
  }
  
  // Parse affected counties
  const affectedCounties = [];
  for (const county of properties.geocode?.SAME || []) {
    affectedCounties.push(county);
  }
  
  return new Alert(
    properties.id || "",
    properties.event || "",
    properties.headline || "",
    properties.description || "",
    properties.instruction,
    properties.severity || "",
    properties.certainty || "",
    properties.urgency || "",
    sent,
    effective,
    onset,
    expires,
    ends,
    properties.status || "",
    properties.messageType || "",
    properties.category || "",
    properties.response || "",
    affectedZones,
    affectedCounties
  );
}

/**
 * Get coordinates for an address using a geocoding service.
 * @param {string} address - Address to geocode.
 * @returns {Coordinates|null} Coordinates for the address, or null if geocoding failed.
 */
function getCoordinatesFromAddress(address) {
  try {
    // This is a placeholder. In a real implementation, you would use a geocoding service
    // like Google Maps, Nominatim, or similar.
    
    // For now, return null to indicate that geocoding is not implemented
    return null;
  } catch (error) {
    return null;
  }
}

/**
 * Convert Celsius to Fahrenheit.
 * @param {number} celsius - Temperature in Celsius.
 * @returns {number} Temperature in Fahrenheit.
 */
function celsiusToFahrenheit(celsius) {
  return (celsius * 9/5) + 32;
}

/**
 * Convert Fahrenheit to Celsius.
 * @param {number} fahrenheit - Temperature in Fahrenheit.
 * @returns {number} Temperature in Celsius.
 */
function fahrenheitToCelsius(fahrenheit) {
  return (fahrenheit - 32) * 5/9;
}

/**
 * Convert meters to miles.
 * @param {number} meters - Distance in meters.
 * @returns {number} Distance in miles.
 */
function metersToMiles(meters) {
  return meters / 1609.344;
}

/**
 * Convert miles to meters.
 * @param {number} miles - Distance in miles.
 * @returns {number} Distance in meters.
 */
function milesToMeters(miles) {
  return miles * 1609.344;
}

/**
 * Convert kilometers per hour to miles per hour.
 * @param {number} kph - Speed in kilometers per hour.
 * @returns {number} Speed in miles per hour.
 */
function kphToMph(kph) {
  return kph * 0.621371;
}

/**
 * Convert miles per hour to kilometers per hour.
 * @param {number} mph - Speed in miles per hour.
 * @returns {number} Speed in kilometers per hour.
 */
function mphToKph(mph) {
  return mph * 1.60934;
}

module.exports = {
  parseForecast,
  parseObservation,
  parseAlert,
  getCoordinatesFromAddress,
  celsiusToFahrenheit,
  fahrenheitToCelsius,
  metersToMiles,
  milesToMeters,
  kphToMph,
  mphToKph
};

File: ruby/nws_utils.rb
require 'date'
require_relative 'weather_models'

module NWSUtils
  # Parse raw forecast data into a Forecast object
  def self.parse_forecast(forecast_data)
    properties = forecast_data["properties"] || {}
    updated_str = properties["updated"] || ""
    
    updated = updated_str.empty? ? nil : DateTime.parse(updated_str.gsub("Z", "+00:00"))
    
    periods = []
    (properties["periods"] || []).each do |period_data|
      start_time = DateTime.parse(period_data["startTime"].gsub("Z", "+00:00"))
      end_time = DateTime.parse(period_data["endTime"].gsub("Z", "+00:00"))
      
      temperature = Temperature.new(
        value: period_data["temperature"] || 0,
        unit: period_data["temperatureUnit"] || "F"
      )
      
      wind_direction = period_data["windDirection"] || ""
      wind_speed_str = period_data["windSpeed"] || "0 mph"
      wind_speed = wind_speed_str.scan(/\d+/).first.to_i
      
      wind = Wind.new(
        speed: wind_speed,
        direction: wind_direction,
        unit: "mph"
      )
      
      precip_prob = period_data.dig("probabilityOfPrecipitation", "value")
      
      period = ForecastPeriod.new(
        name: period_data["name"] || "",
        start_time: start_time,
        end_time: end_time,
        temperature: temperature,
        wind: wind,
        short_forecast: period_data["shortForecast"] || "",
        detailed_forecast: period_data["detailedForecast"] || "",
        icon: period_data["icon"],
        precipitation_probability: precip_prob
      )
      
      periods << period
    end
    
    Forecast.new(updated: updated, periods: periods)
  end

  # Parse raw observation data into an Observation object
  def self.parse_observation(observation_data)
    properties = observation_data["properties"] || {}
    
    station = properties["station"] || ""
    timestamp_str = properties["timestamp"] || ""
    timestamp = timestamp_str.empty? ? DateTime.now : DateTime.parse(timestamp_str.gsub("Z", "+00:00"))
    
    temperature = nil
    if properties.key?("temperature")
      temperature = Temperature.new(
        value: properties["temperature"]["value"] || 0,
        unit: (properties["temperature"]["unitCode"] || "wmoUnit:degC").split(":").last
      )
    end
    
    dewpoint = nil
    if properties.key?("dewpoint")
      dewpoint = Temperature.new(
        value: properties["dewpoint"]["value"] || 0,
        unit: (properties["dewpoint"]["unitCode"] || "wmoUnit:degC").split(":").last
      )
    end
    
    relative_humidity = properties.dig("relativeHumidity", "value")
    
    wind = nil
    if properties.key?("windSpeed") && properties.key?("windDirection")
      wind = Wind.new(
        speed: properties["windSpeed"]["value"] || 0,
        direction: properties["windDirection"]["value"] || 0,
        unit: (properties["windSpeed"]["unitCode"] || "wmoUnit:km_h-1").split(":").last
      )
    end
    
    barometric_pressure = properties.dig("barometricPressure", "value")
    visibility = properties.dig("visibility", "value")
    text_description = properties["textDescription"] || ""
    
    precipitation_last_hour = nil
    if properties.key?("precipitationLastHour")
      precipitation_last_hour = Precipitation.new(
        value: properties["precipitationLastHour"]["value"] || 0,
        unit: (properties["precipitationLastHour"]["unitCode"] || "wmoUnit:mm").split(":").last
      )
    end
    
    Observation.new(
      station: station,
      timestamp: timestamp,
      temperature: temperature,
      dewpoint: dewpoint,
      relative_humidity: relative_humidity,
      wind: wind,
      barometric_pressure: barometric_pressure,
      visibility: visibility,
      text_description: text_description,
      precipitation_last_hour: precipitation_last_hour
    )
  end

  # Parse raw alert data into an Alert object
  def self.parse_alert(alert_data)
    properties = alert_data["properties"] || {}
    
    # Parse timestamps
    sent = DateTime.parse(properties["sent"].gsub("Z", "+00:00"))
    effective = DateTime.parse(properties["effective"].gsub("Z", "+00:00"))
    expires = DateTime.parse(properties["expires"].gsub("Z", "+00:00"))
    
    onset_str = properties["onset"]
    onset = onset_str ? DateTime.parse(onset_str.gsub("Z", "+00:00")) : nil
    
    ends_str = properties["ends"]
    ends = ends_str ? DateTime.parse(ends_str.gsub("Z", "+00:00")) : nil
    
    # Parse affected zones
    affected_zones = []
    (properties["affectedZones"] || []).each do |zone|
      zone_id = zone.split("/").last
      affected_zones << zone_id
    end
    
    # Parse affected counties (derived from geocode)
    affected_counties = []
    (properties.dig("geocode", "SAME") || []).each do |county|
      affected_counties << county
    end
    
    Alert.new(
      id: properties["id"] || "",
      event: properties["event"] || "",
      headline: properties["headline"] || "",
      description: properties["description"] || "",
      instruction: properties["instruction"],
      severity: properties["severity"] || "",
      certainty: properties["certainty"] || "",
      urgency: properties["urgency"] || "",
      sent: sent,
      effective: effective,
      onset: onset,
      expires: expires,
      ends: ends,
      status: properties["status"] || "",
      message_type: properties["messageType"] || "",
      category: properties["category"] || "",
      response_type: properties["response"] || "",
      affected_zones: affected_zones,
      affected_counties: affected_counties
    )
  end

  # Get coordinates for an address using a geocoding service
  def self.get_coordinates_from_address(address)
    begin
      # This is a placeholder. In a real implementation, you would use a geocoding service
      # like Google Maps, Nominatim, or similar.
      nil
    rescue => e
      nil
    end
  end

  # Convert Celsius to Fahrenheit
  def self.celsius_to_fahrenheit(celsius)
    (celsius * 9.0/5.0) + 32
  end

  # Convert Fahrenheit to Celsius
  def self.fahrenheit_to_celsius(fahrenheit)
    (fahrenheit - 32) * 5.0/9.0
  end

  # Convert meters to miles
  def self.meters_to_miles(meters)
    meters / 1609.344
  end

  # Convert miles to meters
  def self.miles_to_meters(miles)
    miles * 1609.344
  end

  # Convert kilometers per hour to miles per hour
  def self.kph_to_mph(kph)
    kph * 0.621371
  end

  # Convert miles per hour to kilometers per hour
  def self.mph_to_kph(mph)
    mph * 1.60934
  end
end
